### PR TITLE
streaming: Handle ACL updates and move connect service health into its own topic

### PR DESCRIPTION
**Pull request by kyhavlov**
This PR is a follow-on from #6310, split out to make it easier to review before the final merge. This PR includes reworking of the event publisher to handle ACL updates in order to cause blocking queries to return immediately with the correct, updated/filtered results when an ACL token/policy/role is changed. This also splits out service health connect events into a separate topic. 

These are the last things that needed to be addressed in order to get the initial cut of the backend streaming work merged.

A follow-up PR adding a custom gRPC resolver can be found at #6625 as well.

The remaining things that need review are: 
#6625 (grpc resolver) from @banks 
#6584 (this PR) for the acl state store changes as per Matt's comments, from @mkeeler and @banks

### CODE REVIEW

**Comment by banks** [comment]
> One last thing to do in this PR is adding a custom grpc Resolver to match the existing server RPC balancing, which I'll push here either later today or Monday.

Suggest we make that a separate PR and work on merging the earlier one and this into the feature branch rather than leave them all open until one big merge happens.

File: agent/consul/streaming_endpoint.go [hunk 1]
```diff
@@ -117,21 +131,30 @@ func sendEvent(event stream.Event, eventFilter stream.EventFilterFunc, aclFilter
 	id := event.ID()
```
Comment (banks) [inline agent/consul/streaming_endpoint.go]: IMO we should do the hashing here with a comment about why and the assumptions about collision probability and safety. etc.

File: agent/agent.go [hunk 2]
```diff
@@ -3918,13 +3918,32 @@ func (a *Agent) registerCache() {
...
+			RefreshTimeout: 10 * time.Minute,
+		})
+		a.cache.RegisterType(cachetype.StreamingHealthServicesConnectName, &cachetype.StreamingHealthServices{
+			Client:  a.streamClient,
+			Connect: true,
+		}, &cache.RegisterOptions{
+			Refresh:        true,
+			RefreshTimer:   0 * time.Second,
+			RefreshTimeout: 10 * time.Minute,
+		})
```
Comment (banks) [inline agent/agent.go]: Why have different cache types for these?

If there is only one cache type and we just use the `structs.ServiceSpecificRequest.Connect` field to switch topic inside the cache type it means less code here and at the call site where we have to make a cached request.

File: agent/cache-types/streaming.go [hunk 3]
```diff
@@ -89,6 +91,21 @@ func (s *Subscriber) run() {
 			return
 		}
 
+		// Check to see if this is a special "reload stream" message before processing
+		// the event.
+		if event.GetReloadStream() {
+			fmt.Printf("[WARN] got ACL reset event\n")
```
Comment (banks) [inline agent/cache-types/streaming.go]: Debug? Probably shouldn't log this at WARN and if we do we should include more context and use a real logger!

File: agent/cache-types/streaming.go [hunk 4]
```diff
@@ -89,6 +91,21 @@ func (s *Subscriber) run() {
...
+			s.handler.Reset()
+
+			// Cancel the context to end the streaming call and create a new one.
+			s.lock.Lock()
+			s.cancelFunc()
+			s.ctx, s.cancelFunc = context.WithCancel(context.Background())
+			s.lock.Unlock()
+
+			goto START
```
Comment (banks) [inline agent/cache-types/streaming.go]: Jumping back to start with no backoff makes me wonder if there are error cases where this can get into a hot loop.

I don't _think_ it should in normal use but I wonder if we should have some kind of exponential backoff on this just incase we somehow get into a loop because of a server bug or something?

File: agent/cache-types/streaming_health_services.go [hunk 5]
```diff
@@ -30,13 +32,16 @@ func (c *StreamingHealthServices) Fetch(opts cache.FetchOptions, req cache.Reque
...
 		Index:  reqReal.MinQueryIndex,
 		Filter: reqReal.Filter,
-		TopicFilters: &stream.Filters{
-			Connect: reqReal.Connect,
-			Tags:    reqReal.ServiceTags,
-		},
 	}
+
+	// Switch the topic if Connect is enabled.
+	if c.Connect {
```
Comment (banks) [inline agent/cache-types/streaming_health_services.go]: per comment above this could just switch to the request:

```suggestion
	if reqReal.Connect {
```

File: agent/cache-types/streaming_health_services.go [hunk 6]
```diff
@@ -81,6 +86,11 @@ func (h *healthServicesHandler) State(idx uint64) interface{} {
 	for _, node := range h.state {
 		result.Nodes = append(result.Nodes, node)
 	}
+	result.Nodes.Sort()
 	result.Index = idx
 	return &result
 }
+
+func (h *healthServicesHandler) Reset() {
+	h.state = make(map[string]structs.CheckServiceNode)
```
Comment (banks) [inline agent/cache-types/streaming_health_services.go]: Think it might be worth comment to point out that this isn't actually wiping the result in the cache - until we get a new full snapshot we won't server clients with an empty or partial request but the last good one from before the reset.

That is correct right?

File: agent/cache-types/streaming.go [hunk 7]
```diff
@@ -168,6 +185,8 @@ WAIT:
 		}
 
 		// Wait for another update if the result wasn't higher than the requested index.
+		// If this update came from a snapshot finishing, we return no matter what, as the
+		// state/index could have changed as a result of ACL updates or other filtering.
```
[the remark this answers is not in the document]
Comment (kyhavlov) [reply agent/cache-types/streaming.go]: I think this is a leftover comment that can be removed, I had something here to handle the snapshot finishing before sending the max index with the "end of snapshot" message

File: agent/consul/state/acl.go [hunk 8]
```diff
@@ -986,6 +1027,11 @@ func (s *Store) aclTokenGetTxn(tx *memdb.Txn, ws memdb.WatchSet, value, index st
 
 	if rawToken != nil {
 		token := rawToken.(*structs.ACLToken)
+
+		if !fixLinks {
+			return token, nil
+		}
+
```
[the remark this answers is not in the document]
Comment (kyhavlov) [reply agent/consul/state/acl.go]: In the event publisher when a policy/role gets deleted we look up the subscriber tokens to see which ones are affected, and if the links get fixed in the process then we can't tell which ones had the policy/role that was deleted and need to be notified

File: agent/consul/state/event.go [hunk 9]
```diff
@@ -37,13 +63,47 @@ func (e *EventPublisher) PreparePublish(events []stream.Event) error {
...
 }
 
-// Commit sends any staged events to the relevant listeners. This is called
+// Commit triggers any staged events to be sent to the relevant listeners. This is called
 // via txn.Defer to delay it from running until the transaction has been finalized.
 func (e *EventPublisher) Commit() {
+	e.stagedLock.Lock()
+	defer e.stagedLock.Unlock()
+
+	e.commitCh <- commitUpdate{
```
[the remark this answers is not in the document]
Comment (kyhavlov) [reply agent/consul/state/event.go]: This was to remove the event publishing work from the hot path in case the publishing/acl update logic ever adds a significant amount of time to commits. We talked about doing this to avoid raft commits potentially taking too long

File: agent/agent.go (hunk 2 above)
Comment (kyhavlov) [reply agent/agent.go]: I didn't realize the Connect field was part of the cache key, that's definitely cleaner.

File: agent/cache-types/streaming_health_services.go (hunk 6 above)
Comment (kyhavlov) [reply agent/cache-types/streaming_health_services.go]: Yes that's accurate, do you think we should be resetting the cache entry here too?

File: agent/cache-types/streaming.go (hunk 4 above)
Comment (kyhavlov) [reply agent/cache-types/streaming.go]: I think without some sort of event buffer to keep the acl update events around in the publisher we couldn't get stuck in a hot loop here since that's the only source of the reload event - that said, I agree it's probably better to add a backoff here now to avoid something unexpected.

File: agent/consul/state/event.go [hunk 10]
```diff
@@ -65,8 +125,73 @@ func (e *EventPublisher) Commit() {
...
 		}
 	}
+}
 
-	e.staged = nil
+func (e *EventPublisher) handleACLUpdate(tx *memdb.Txn, event stream.Event) error {
+	switch event.Topic {
+	case stream.Topic_ACLTokens:
+		token := event.GetACLToken()
+		subscribers, ok := e.tokenSubs[token.Token.SecretID]
```
[the remark this answers is not in the document]
Comment (kyhavlov) [reply agent/consul/state/event.go]: Accessor seems like a better fit here, it's useful to be able to iterate through the subscribers' tokens when a policy/role changes and a hash wouldn't let us do that

[the remark this answers is not in the document]
Comment (kyhavlov) [reply agent/consul/state/event.go]: I rolled this back after realizing this wouldn't work with legacy ACLs because those tokens don't have a separate accessor - we could use a hash but then we'd have to iterate through every token in the store instead of just the ones used by subscribers. What do you think?

File: agent/cache-types/streaming_health_services.go (hunk 6 above)
Comment (banks) [reply agent/cache-types/streaming_health_services.go]: Nope, keeping "last good result" in cache is important for the "fail static" story. Cache is inherently stale already so an older usable response is better than nothing for this stuff for sure.

File: agent/consul/state/event.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (banks) [reply agent/consul/state/event.go]: @mkeeler is that true? I think we auto-gen accessor ids for legacy ACLs on startup with 1.4 but I forget maybe we didn't do that or there is an interim where they might be empty.

Honestly It doesn't change our threat model so it's probably OK as it is, I just dislike the idea that secrets are used as keys in places they don't really need to be. In practice I _think_ Go will not even make a copy of the string contents here and it's really just another pointer to the same string we already have in memory so it's not even that, it's more that it would be easy to debug-print the map or something later without thinking through and end up risking exposing secrets because it wasn't obvious that the component you were debugging was sensitive.

tl;dr I'm _OK_ leaving this for simplicity but wish we could use accessor more for stuff like this.

 > it's useful to be able to iterate through the subscribers' tokens when a policy/role changes and a hash wouldn't let us do that

Maybe I'm missing something but surely you can iterate through subscribers tokens, hash each one and then lookup? I'm not sure sure it's warranted thought per above.

File: agent/consul/state/acl.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (banks) [reply agent/consul/state/acl.go]: 🤔 got it.

I'm not familiar enough with how and when this fixup process works to know if this is right/safe or not. Since we added the channel into event publish to make it Async, that now means that this read in response to an ACL DELETE event is not atomic with the DELETE itself - if there is some background process that is cleaning up the links for example it could be possible that that process picked up the delete even quicker and performed the actual fixup on the memDB snapshot that we are now querying.

In other words, whatever was broken before you added this, could still be broken the same way in a rare race which makes me think we should maybe have a different approach.

[edit]

Actually I was about to propose it would be safer if we pass through a snapshot on the `commitCh` so we know that any reads done during publish are effectively "atomic" with the state at the publish time but then I checked and you already do that.

So I *think* this is OK but still not 100% sure why it's the right thing to do given that fixing up links is probably something we should ideally not make async and if we _did_ do them inline it would be equivalent of this change not being present. Do we just need need to read the ACL token before the delete and somehow pass that through to the event rather than rely on us not getting the state store referentially consistent right after the delete?

[the remark this answers is not in the document]
Comment (mkeeler) [reply agent/consul/state/acl.go]: I think this makes sense. Also if you are not sending the retrieved token/role back to the user then there shouldn't be any need to fix the links at all. All of the internal ACL token resolution related things take into account linking to deleted things and nothing internally cares about the link's Name.

[the remark this answers is not in the document]
Comment (kyhavlov) [reply agent/consul/state/acl.go]: @mkeeler so you're saying this change (skipping link fixing) shouldn't be necessary? I only added it after hitting test failures where looking up the token was removing the deleted policy/role from the links

File: agent/consul/streaming_endpoint.go [hunk 11]
```diff
@@ -47,23 +47,33 @@ func (h *ConsulGRPCAdapter) Subscribe(req *stream.SubscribeRequest, server strea
...
-			if err := sendEvent(event, eventFilter, aclFilter, eval, sent, server); err != nil {
+			if err := sendEvent(event, aclFilter, eval, sent, server); err != nil {
 				return err
 			}
 		}
-	}
+	} else {
+		// If there wasn't a snapshot, just send an end of snapshot message
+		// so the client knows not to wait for one.
+		idx, err := state.ComputeIndex()
```
[the remark this answers is not in the document]
Comment (kyhavlov) [reply agent/consul/streaming_endpoint.go]: I think you're right, it should be the same index.

File: agent/consul/state/event.go [hunk 12]
```diff
@@ -65,8 +125,98 @@ func (e *EventPublisher) Commit() {
...
+		// If there are subscribers using the updated/deleted token, signal them
+		// to reload their connection.
+		if ok {
+			e.reloadSubscribers(subscribers)
+		}
+	case stream.Topic_ACLPolicies:
+		policy := event.GetACLPolicy()
+		affectedSubs := make(map[*stream.SubscribeRequest]chan stream.Event)
+		for token, subscribers := range e.tokenSubs {
+			token, err := e.store.aclTokenGetTxnFixupLinks(tx, nil, token, "id", false)
```
[the remark this answers is not in the document]
Comment (kyhavlov) [reply agent/consul/state/event.go]: I totally missed that ACLTokenList looks up tokens by policy used. That seems better than what I was doing here + it looks like we don't need the logic for skipping link fixing in that case.

File: agent/consul/state/acl.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (banks) [reply agent/consul/state/acl.go]: I think Matt was saying that your code here makes sense rather than my rambling pseudo concern @kyhavlov  😄

File: agent/consul/state/acl.go [hunk 13]
```diff
@@ -785,6 +788,14 @@ func (s *Store) ACLTokenSet(idx uint64, token *structs.ACLToken, legacy bool) er
 	tx := s.db.Txn(true)
 	defer tx.Abort()
 
+	existing, err := s.aclTokenGetTxn(tx, nil, token.SecretID, "id")
+	if err != nil {
+		return err
+	}
+	if existing != nil {
+		existing.SetHash(false)
```
[the remark this answers is not in the document]
Comment (kyhavlov) [reply agent/consul/state/acl.go]: After your comment above I realized I was only hitting the hash not being set during state store tests, so I removed this and just set the hash manually during the tests to simulate the endpoints setting it.

File: agent/consul/state/event.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (banks) [reply agent/consul/state/event.go]: What was the final conclusion @kyhavlov? I'm OK with leaving it for now I guess but it seems there are at least a couple of options here (hashing or dissallowing non-upgraded legacy tokens) that would be easy enough to do?

File: agent/consul/state/event.go [hunk 14]
```diff
@@ -65,8 +125,98 @@ func (e *EventPublisher) Commit() {
...
+
+			// If the updated policy was used in this token, add any subscribers using it
+			// to the affected set.
+			for _, p := range token.Policies {
+				if p.ID == policy.PolicyID {
+					for sub, ch := range subscribers {
+						affectedSubs[sub] = ch
+					}
+					break
+				}
```
[the remark this answers is not in the document]
Comment (banks) [reply agent/consul/state/event.go]: @kyhavlov was this done?

File: agent/consul/state/acl.go [hunk 15]
```diff
@@ -978,6 +1009,10 @@ func (s *Store) ACLTokenBatchGet(ws memdb.WatchSet, accessors []string) (uint64,
 }
 
 func (s *Store) aclTokenGetTxn(tx *memdb.Txn, ws memdb.WatchSet, value, index string) (*structs.ACLToken, error) {
+	return s.aclTokenGetTxnFixupLinks(tx, ws, value, index, true)
+}
+
+func (s *Store) aclTokenGetTxnFixupLinks(tx *memdb.Txn, ws memdb.WatchSet, value, index string, fixLinks bool) (*structs.ACLToken, error) {
```
Comment (banks) [inline agent/consul/state/acl.go]: Do we still need both variants? I think you said that we don't actually need to avoid the fixup any more right?

File: agent/consul/state/event.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (kyhavlov) [reply agent/consul/state/event.go]: Currently we're storing hashes of the secret IDs to track changes. I don't have a strong preference either way but it seems nice to be backwards compatible by using the secret. If we had to switch over to accessors in the future it wouldn't be a large change since this is purely internal and ephemeral (recreated on server restart), so it we officially stop supporting legacy ACLs at some point this wouldn't be a point of friction.

File: agent/consul/state/event.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (kyhavlov) [reply agent/consul/state/event.go]: Yes, right below this comment we look up the roles using this policy and update any subscribers using tokens with those roles.

File: agent/consul/streaming_endpoint.go [hunk 16]
```diff
@@ -128,11 +128,8 @@ func sendEvent(event stream.Event, aclFilter *aclFilter, eval *bexpr.Evaluator,
 	// hash it to save space. This is only used to determine if an object has been
 	// removed and needs to be deleted from the client's cache so we make the tradeoff
 	// of using less memory here at the extremely small risk of a hash collision.
-	rawId := event.ID()
-	hash, err := blake2b.New256(nil)
-	if err != nil {
-		return err
-	}
+	rawId := event.ContentID()
```
Comment (banks) [inline agent/consul/streaming_endpoint.go]: I suggest for the comment above that we make it a little more precise. Something like:

```
	// Get the unique identifier for the object the event pertains to, and
	// hash it to save space. This is only used to determine if an object has been
	// removed and needs to be deleted from the client's cache. In other words, if we hit a
        // hash collision, it only results in a redundant message being sent without affecting 
        // correctness. 32 bits means there is only a 1% chance of collision with ~9000 items in this map.
        // For current uses that seems reasonable but when we add KV streaming we may want
        // to revisit this.
```

File: agent/consul/streaming_endpoint.go [hunk 17]
```diff
@@ -128,11 +128,8 @@ func sendEvent(event stream.Event, aclFilter *aclFilter, eval *bexpr.Evaluator,
...
+	hash := fnv.New32a()
 	hash.Write([]byte(rawId))
 	bytes := hash.Sum(nil)
 	idHash := string(bytes)
```
Comment (banks) [inline agent/consul/streaming_endpoint.go]: If we are going to the bother of a 32 bit hash to save space it's wasteful to convert it back into a string again and have the overhead of another string allocation and string map (which implies pointers and size counters per key etc). This string will also be invalid UTF-8 since the bytes are essentially random.

We should make the map be keyed by `uint32` and then use `hash.Uint32()` to generate the key.

**Comment by banks** [comment]
Approved on the basis that the hash fix from the previous PR is fixed and commented as mentioned above in https://github.com/hashicorp/consul/pull/6584#pullrequestreview-303998778.

I'm approving ahead of time because that is a trivial fix and I'm going to be out for a while. Provided that is done this is good to be merged.

Great job Kyle!
